plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange/ASTPreProcess.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ide.Plugin.SelectionRange.ASTPreProcess
+    ( preProcessAST
+    , PreProcessEnv(..)
+    ) where
+
+import           Control.Monad.Reader            (Reader, asks)
+import           Data.Foldable                   (find, foldl')
+import           Data.List                       (groupBy)
+import qualified Data.Map.Strict                 as Map
+import           Data.Maybe                      (mapMaybe)
+import qualified Data.Set                        as Set
+import           Development.IDE.GHC.Compat      (ContextInfo (MatchBind, TyDecl, ValBind),
+                                                  HieAST (..), Identifier,
+                                                  IdentifierDetails (identInfo),
+                                                  NodeInfo (NodeInfo, nodeIdentifiers),
+                                                  RefMap, Span, flattenAst,
+                                                  isAnnotationInNodeInfo,
+                                                  mkAstNode, mkRealSrcSpan,
+                                                  nodeInfoFromSource,
+                                                  realSrcSpanEnd,
+                                                  realSrcSpanStart)
+import           Development.IDE.GHC.Compat.Util (FastString)
+import           Prelude                         hiding (span)
+
+newtype PreProcessEnv a = PreProcessEnv
+    { preProcessEnvRefMap :: RefMap a
+    }
+
+-- | Make the AST more suitable for generating selection range.
+preProcessAST :: HieAST a -> Reader (PreProcessEnv a) (HieAST a)
+preProcessAST node = mergeImports node >>= mergeSignatureWithDefinition
+
+mergeImports :: HieAST a -> Reader (PreProcessEnv a) (HieAST a)
+mergeImports node = pure $ node { nodeChildren = children }
+  where
+    children = mapMaybe merge
+        . groupBy (\x y -> nodeIsImport x && nodeIsImport y)
+        . nodeChildren $ node
+
+    merge []  = Nothing
+    merge [x] = Just x
+    merge xs  = Just (createVirtualNode xs)
+
+nodeIsImport :: HieAST a -> Bool
+nodeIsImport = isAnnotationInAstNode ("ImportDecl", "ImportDecl")
+
+createVirtualNode :: [HieAST a] -> HieAST a
+createVirtualNode nodes = mkAstNode (NodeInfo mempty mempty mempty) span' nodes
+  where
+    span' = mkRealSrcSpan (minimum locations) (maximum locations)
```

Comment:
https://hackage.haskell.org/package/ghc-9.2.1/docs/GHC-Types-SrcLoc.html#v:combineRealSrcSpans

Reply:
Maybe you can just fold up the source spans with that.

Reply:
oh, this is really helpful